--- a/packages/gnu/targets/linkcmd.xdt	2010-02-25 14:23:03.000000000 -0500
+++ b/packages/gnu/targets/linkcmd.xdt	2011-02-16 20:09:58.000000000 -0500
@@ -53,22 +53,4 @@
 %}
 
 %var _utils = xdc.loadCapsule("gnu/targets/linkUtils.xs");
-%/* Mingw compilers do not work with the SECTIONS directive.
-% */
-%if (prog.build.target.$name != "gnu.targets.Mingw") {
-SECTIONS {
-`_utils.genSections(prog)`
-}
 
-%if (xdc.om['xdc.runtime.Text'].isLoaded != true) {
-ASSERT(ADDR(xdc.noload) + SIZEOF(xdc.noload) < ADDR(.rodata)
-    || ADDR(.rodata) + SIZEOF(.rodata) < ADDR(xdc.noload),
-    "Sections .rodata and xdc.noload overlap. See xdc.runtime.Text.isLoaded for more.")
-%}
-%} // end of the workaround for Mingw
-%%{
-/*
- * @(#) gnu.targets; 1, 0, 1,387; 2-24-2010 16:24:01; /db/ztree/library/trees/xdctargets/xdctargets-b36x/src/
- */
-
-%%}
