--- a/packages/ti/sdo/dmai/linux/dm6446/Display_fbdev.c	2011-11-28 10:42:25.000000000 -0500
+++ b/packages/ti/sdo/dmai/linux/dm6446/Display_fbdev.c	2011-11-28 11:31:12.000000000 -0500
@@ -138,7 +138,6 @@
     Int                      displaySize;
     Int                      bufIdx;
     Int8                    *virtPtr;
-    Int32                    physPtr;
     Int32                    height, width;
     Display_Handle           hDisplay;
     Buffer_Handle            hBuf;
@@ -272,9 +271,6 @@
 
     _Dmai_blackFill(hBuf);
 
-    Dmai_dbg3("Display buffer %d mapped to %#lx has physical address %#lx\n", 0,
-              (Int32) virtPtr, physPtr);
-
     for (bufIdx=1; bufIdx < attrs->numBufs; bufIdx++) {
         hBuf = BufTab_getBuf(hBufTab, bufIdx);
         Buffer_setNumBytesUsed(hBuf, varInfo.xres * varInfo.yres *
@@ -283,9 +279,6 @@
         virtPtr = virtPtr + displaySize;
         Buffer_setUserPtr(hBuf, virtPtr);
         _Dmai_blackFill(hBuf);
-
-        Dmai_dbg3("Display buffer %d mapped to %#lx, physical address %#lx\n",
-                  bufIdx, (unsigned long) virtPtr, physPtr);
     }
 
     hDisplay->hBufTab = hBufTab;
