Index: ti-codecs/dm6446_dvsdk_combos_2_05/packages/ti/sdo/servers/decode/decode.cfg
===================================================================
--- ti-codecs.orig/dm6446_dvsdk_combos_2_05/packages/ti/sdo/servers/decode/decode.cfg	2012-02-07 17:09:51.538023111 -0800
+++ ti-codecs/dm6446_dvsdk_combos_2_05/packages/ti/sdo/servers/decode/decode.cfg	2012-02-07 17:14:22.911368784 -0800
@@ -31,11 +31,11 @@
     print("Configuring server to use watermarked codecs...");
     
 
-    H264DEC.alg.watermark = true;
-    MPEG2DEC.alg.watermark = true;
-    MPEG4DEC.alg.watermark = true;
-    AACHEDEC.alg.watermark = true;    
-    G711DEC.alg.watermark = true;
+    H264DEC.alg.watermark = false;
+    MPEG2DEC.alg.watermark = false;
+    MPEG4DEC.alg.watermark = false;
+    AACHEDEC.alg.watermark = false;    
+    G711DEC.alg.watermark = false;
 
 }
 else
Index: ti-codecs/dm6446_dvsdk_combos_2_05/packages/ti/sdo/servers/encode/encode.cfg
===================================================================
--- ti-codecs.orig/dm6446_dvsdk_combos_2_05/packages/ti/sdo/servers/encode/encode.cfg	2012-02-07 17:06:50.753126656 -0800
+++ ti-codecs/dm6446_dvsdk_combos_2_05/packages/ti/sdo/servers/encode/encode.cfg	2012-02-07 17:14:22.891368684 -0800
@@ -28,9 +28,9 @@
     /* This server is for evaluation, so use the watermarked versions */
     print("Configuring server to use watermarked codecs...");
     
-    H264ENC.alg.watermark = true;    
-    MPEG4ENC.alg.watermark = true;
-    G711ENC.alg.watermark = true;
+    H264ENC.alg.watermark = false;    
+    MPEG4ENC.alg.watermark = false;
+    G711ENC.alg.watermark = false;
     
 }
 else
Index: ti-codecs/dm6446_dvsdk_combos_2_05/packages/ti/sdo/servers/loopback/loopback.cfg
===================================================================
--- ti-codecs.orig/dm6446_dvsdk_combos_2_05/packages/ti/sdo/servers/loopback/loopback.cfg	2012-02-07 17:10:16.038144604 -0800
+++ ti-codecs/dm6446_dvsdk_combos_2_05/packages/ti/sdo/servers/loopback/loopback.cfg	2012-02-07 17:10:22.722177752 -0800
@@ -25,8 +25,8 @@
 if (prog.build.cfgArgs.profile == "eval") {
     /* This server is for evaluation, so use the watermarked versions */
     print("Configuring server to use watermarked codecs...");
-    H264ENC.alg.watermark = true;
-    H264DEC.alg.watermark = true;
+    H264ENC.alg.watermark = false;
+    H264DEC.alg.watermark = false;
 }
 else
 {
