Index: trunk/include/download.mk
===================================================================
--- trunk/include/download.mk	(revision 42297)
+++ trunk/include/download.mk	(working copy)
@@ -46,7 +46,7 @@
 endef
 
 define wrap_mirror
-	$(if $(if $(MIRROR),$(filter-out x,$(MIRROR_MD5SUM))),@$(SCRIPT_DIR)/download.pl "$(DL_DIR)" "$(FILE)" "$(MIRROR_MD5SUM)" || ( $(1) ),$(1))
+	$(if $(if $(MIRROR),$(filter-out x,$(MIRROR_MD5SUM))),@$(SCRIPT_DIR)/download.pl "$(DL_DIR)" "$(FILE)" "$(MIRROR_MD5SUM)" || ( $(1) ), ( $(1) ) || $(DownloadMethod/default))
 endef
 
 define DownloadMethod/cvs
