Index: trunk/feeds.conf.default
===================================================================
--- trunk/feeds.conf.default	(revision 42297)
+++ trunk/feeds.conf.default	(working copy)
@@ -1,9 +1,12 @@
+# Generic Gateworks package feed
+src-svn gateworks http://svn.gateworks.com/openwrt/trunk/packages
+# Upstream OpenWRT package feeds
 src-git packages https://github.com/openwrt/packages.git
 src-git luci http://git.openwrt.org/project/luci.git
 src-git routing https://github.com/openwrt-routing/packages.git
 src-git telephony http://git.openwrt.org/feed/telephony.git
 src-git management https://github.com/openwrt-management/packages.git
-#src-git oldpackages http://git.openwrt.org/packages.git
+src-git oldpackages http://git.openwrt.org/packages.git
 #src-svn xwrt http://x-wrt.googlecode.com/svn/trunk/package
 #src-svn phone svn://svn.openwrt.org/openwrt/feeds/phone
 #src-svn efl svn://svn.openwrt.org/openwrt/feeds/efl
