--- a/packages/gnu/targets/arm/linkcmd.xdt	2010-02-25 14:23:03.000000000 -0500
+++ b/packages/gnu/targets/arm/linkcmd.xdt	2011-02-16 20:10:08.000000000 -0500
@@ -39,24 +39,3 @@
 %}
 
 %var _utils = xdc.loadCapsule("gnu/targets/linkUtils.xs");
-%/* Bugzilla 295146 prevent us from using the SECTIONS directive for
-% * CodeSourcery target 4.3.3 or newer version.
-% */
-%if (prog.build.target.GCCVERS < "4.3.3"
-%    || prog.build.target.GCCTARG != "arm-none-linux-gnueabi") {
-SECTIONS {
-`_utils.genSections(prog)`
-}
-
-%if (xdc.om['xdc.runtime.Text'].isLoaded != true) {
-ASSERT(ADDR(xdc.noload) + SIZEOF(xdc.noload) < ADDR(.rodata)
-    || ADDR(.rodata) + SIZEOF(.rodata) < ADDR(xdc.noload),
-    "Sections .rodata and xdc.noload overlap. See xdc.runtime.Text.isLoaded for more.")
-%}
-%} // end of the workaround for 295146
-%%{
-/*
- * @(#) gnu.targets.arm; 1, 0, 0, 0,216; 2-24-2010 16:24:01; /db/ztree/library/trees/xdctargets/xdctargets-b36x/src/
- */
-
-%%}
