Index: gstreamer-ti-svn-r611/Makefile.external
===================================================================
--- gstreamer-ti-svn-r611.orig/Makefile.external	2012-03-06 21:33:55.249301126 -0800
+++ gstreamer-ti-svn-r611/Makefile.external	2012-03-06 21:34:32.801487342 -0800
@@ -47,7 +47,7 @@
 endif
 
 ifeq ($(PLATFORM), dm6446)
-    export XDC_TARGET   = gnu.targets.MVArm9
+    export XDC_TARGET   = gnu.targets.UCArm9
     export XDC_PLATFORM = ti.platforms.evmDM6446
 endif
 
@@ -75,7 +75,7 @@
 	./autogen.sh --noconfigure
 
 Makefile: configure
-	./configure CPPFLAGS="$(CPPFLAGS)" --host=$(HOST)
+	./configure CPPFLAGS="$(CPPFLAGS)" --build=i686-linux --host=arm-openwrt-linux --prefix=/usr
 
 omap3530: Makefile
 	$(MAKE) -f Makefile
Index: gstreamer-ti-svn-r611/src/Makefile.am
===================================================================
--- gstreamer-ti-svn-r611.orig/src/Makefile.am  2012-03-06 21:49:17.193872807 -0800
+++ gstreamer-ti-svn-r611/src/Makefile.am       2012-03-06 21:49:42.986000703 -0800
@@ -18,7 +18,7 @@
 # XDC Configuration
 CONFIGURO     = $(XDC_INSTALL_DIR)/xs xdc.tools.configuro
 BUILT_SOURCES = $(XDC_CONFIG_BASENAME)/compiler.opt $(XDC_CONFIG_BASENAME)/linker.cmd
-XDC_PATH      = .;$(XDC_INSTALL_DIR)/packages;$(LINK_XDC_ROOT);$(FC_INSTALL_DIR)/packages;$(CE_INSTALL_DIR)/packages;$(XDAIS_INSTALL_DIR)/packages;$(CODEC_INSTALL_DIR)/packages;$(CMEM_INSTALL_DIR)/packages;$(DMAI_INSTALL_DIR)/packages;$(LPM_INSTALL_DIR)/packages;$(XDC_USER_PATH);$(EDMA3_LLD_INSTALL_DIR)/packages;
+XDC_PATH      = .;$(XDC_INSTALL_DIR)/packages;$(LINK_INSTALL_DIR)/packages;$(FC_INSTALL_DIR)/packages;$(CE_INSTALL_DIR)/packages;$(XDAIS_INSTALL_DIR)/packages;$(CODEC_INSTALL_DIR)/packages;$(CMEM_INSTALL_DIR)/packages;$(DMAI_INSTALL_DIR)/packages;$(LPM_INSTALL_DIR)/packages;$(XDC_USER_PATH);$(EDMA3_LLD_INSTALL_DIR)/packages;
 
 gstticodecs_platform.c :
        ln -s $(srcdir)/gstticodecs_$(PLATFORM).c gstticodecs_platform.c

