--- a/Makefile.external	2010-08-24 16:20:05.000000000 -0400
+++ b/Makefile.external	2011-05-23 16:29:26.000000000 -0400
@@ -52,8 +52,9 @@
 endif
 
 ifeq ($(GST_TI_PLATFORM), dm6446)
-    export XDC_TARGET   = gnu.targets.MVArm9
+    export XDC_TARGET   = gnu.targets.UCArm9
     export XDC_PLATFORM = ti.platforms.evmDM6446
+    export PLATFORM_XDC  = ${XDC_PLATFORM}
 endif
 
 ifeq ($(GST_TI_PLATFORM), dm6467)
@@ -96,7 +97,7 @@
 	./autogen.sh --noconfigure
 
 Makefile: configure
-	./configure CPPFLAGS="$(CPPFLAGS)" --host=$(HOST)
+	./configure CPPFLAGS="$(CPPFLAGS)" --build=i686-linux --host=arm-openwrt-linux --prefix=/usr
 
 omap3530: Makefile
 	$(MAKE) -f Makefile
