--- a/packages/gnu/targets/UCArm9.xdc	2010-02-25 14:23:03.000000000 -0500
+++ b/packages/gnu/targets/UCArm9.xdc	2011-05-12 14:04:39.000000000 -0400
@@ -32,7 +32,7 @@
     override readonly config string rts = "gnu.targets.rts470uC";
     override config string platform     = "ti.platforms.evmDM6446";
     
-    override config string LONGNAME = "/bin/arm-linux-gcc";
+    override config string LONGNAME = "/bin/arm-openwrt-linux-gcc";
 
     override config ITarget.Options ccOpts = {
         prefix: "-Wunused",
