--- a/dm6446_dvsdk_combos_2_05/packages/ti/sdo/servers/encode/encode.cfg
+++ b/dm6446_dvsdk_combos_2_05/packages/ti/sdo/servers/encode/encode.cfg
@@ -57,6 +57,13 @@ MPEG4ENC.alg.udataSection  = "DDR2";
 H264ENC.manageOutBufsCache  = [ false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false ];
 MPEG4ENC.manageOutBufsCache = [ false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false ];
 
+var AACHEENC = xdc.useModule('ti.sdo.codecs.aacheenc.ce.AACHEENC');
+// Module Config
+	AACHEENC.alg.watermark = false;
+	AACHEENC.alg.codeSection = "DDR2";
+	AACHEENC.alg.dataSection = "DDR2";
+	AACHEENC.alg.udataSection = "DDR2";
+	
 
 Server.algs = [
 
@@ -72,8 +79,10 @@ Server.algs = [
         stackMemId: 0, priority: Server.MINPRI + 2}
     },
 
+    {name: "aacheenc", mod: AACHEENC, groupId: 1, threadAttrs: {
+        stackMemId: 0, priority: Server.MINPRI + 2}
+    },
 ];
-
    /*
     *  ======== DSKT2 and RMAN Configuration ========
     *  XDAIS Algorithm Memory and DMA allocation
