--- a/dm6446_dvsdk_combos_2_05/packages/ti/sdo/servers/decode/decode.cfg
+++ b/dm6446_dvsdk_combos_2_05/packages/ti/sdo/servers/decode/decode.cfg
@@ -60,6 +60,12 @@ MPEG4DEC.alg.codeSection = "DDR2";
 MPEG2DEC.alg.codeSection = "DDR2";
 AACHEDEC.alg.codeSection = "DDR2";
 
+var MP3DEC = xdc.useModule('ti.sdo.codecs.mp3dec.ce.MP3DEC');
+// Module Config
+	MP3DEC.alg.watermark = false;
+	MP3DEC.alg.codeSection = "DDR2";
+	MP3DEC.alg.dataSection = "DDR2";
+	MP3DEC.alg.udataSection = "DDR2";
 
 H264DEC.manageOutBufsCache  = [ false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false ];
 MPEG2DEC.manageOutBufsCache = [ false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false ];
@@ -88,6 +94,9 @@ Server.algs = [
         stackMemId: 0,priority: Server.MINPRI + 2}
     },
 
+    {name: "mp3dec", mod: MP3DEC, groupId: 1, threadAttrs: {
+        stackMemId: 0,priority: Server.MINPRI + 2}
+    },
     
 ];
 
