--- a/src/gstticodecs_dm6446.c
+++ b/src/gstticodecs_dm6446.c
@@ -53,6 +53,10 @@ GstTICodec gst_ticodec_codecs[] = {
         "MPEG1L3 Audio Decoder", /* String name of codec used by plugin      */
         "mp3dec",                /* String name of codec used by CE          */
         decodeEngine             /* Engine that contains this codec          */
+    }, {
+        "AAC Audio Encoder",     /* String name of codec used by plugin      */
+        "aacheenc",              /* String name of codec used by CE          */
+        encodeEngine             /* Engine that contains this codec          */
     },
 
     /* Video Codecs */
