Index: trunk/feeds.conf.default
===================================================================
--- trunk/feeds.conf.default	(revision 36311)
+++ trunk/feeds.conf.default	(working copy)
@@ -1,6 +1,9 @@
+# Generic Gateworks package feed
+src-svn gateworks http://svn.gateworks.com/openwrt/trunk/packages
+# Upstream OpenWRT package feeds
 src-svn packages svn://svn.openwrt.org/openwrt/packages
 src-svn xwrt http://x-wrt.googlecode.com/svn/trunk/package
-src-git luci git://nbd.name/luci.git
+src-svn luci http://svn.luci.subsignal.org/luci/trunk
 #src-svn phone svn://svn.openwrt.org/openwrt/feeds/phone
 #src-svn efl svn://svn.openwrt.org/openwrt/feeds/efl
 #src-svn xorg svn://svn.openwrt.org/openwrt/feeds/xorg
