--- a/packages/config.bld	2011-11-28 10:42:25.000000000 -0500
+++ b/packages/config.bld	2011-11-28 11:24:44.000000000 -0500
@@ -68,10 +68,10 @@
 }
 
 /* location of the GCC Arm9 tools */
-var GCArmv5T = xdc.useModule('gnu.targets.arm.GCArmv5T');
-GCArmv5T.LONGNAME = longName;
-GCArmv5T.platform = java.lang.System.getenv("PLATFORM_XDC");
-GCArmv5T.rootDir = tooldir;
+var UCArm9 = xdc.useModule('gnu.targets.UCArm9');
+UCArm9.LONGNAME = longName;
+UCArm9.platform = java.lang.System.getenv("PLATFORM_XDC");
+UCArm9.rootDir = tooldir;
 
 Build.targets = [
 ];
