--- a/packages/ti/sdo/dmai/linux/dm6446/Framecopy_accel.c	2011-11-28 10:42:25.000000000 -0500
+++ b/packages/ti/sdo/dmai/linux/dm6446/Framecopy_accel.c	2011-11-28 11:36:11.000000000 -0500
@@ -46,7 +46,7 @@
 #include "../../priv/_Framecopy.h"
 
 #ifdef CONFIG_RESIZER
-#include <linux/davinci_resizer.h>
+#include <media/davinci/davinci_resizer.h>
 #endif
 
 #define MODULE_NAME     "Framecopy"
@@ -105,7 +105,7 @@
 #ifdef CONFIG_RESIZER
     Int width, height;
     BufferGfx_Dimensions srcDim, dstDim;
-    struct rsz_params params = {
+    struct rsz_params_t params = {
         0,                              /* in_hsize (set at run time) */
         0,                              /* in_vsize (set at run time) */
         0,                              /* in_pitch (set at run time) */
@@ -218,7 +218,7 @@
                             Buffer_Handle hSrcBuf, Buffer_Handle hDstBuf)
 {
 #ifdef CONFIG_RESIZER
-    struct rsz_resize   rsz;
+    struct rsz_resize_t   rsz;
     Int                 rszError;
     UInt32              srcOffset, dstOffset;
 
