--- a/Platform.make	2011-11-28 10:42:25.000000000 -0500
+++ b/Platform.make	2011-11-28 11:56:35.000000000 -0500
@@ -135,7 +135,7 @@
 PLATFORM_OS_dm6446_al       = linux
 PLATFORM_DEVICE_dm6446_al   = dm6446
 PLATFORM_PERIPHERALS_dm6446_al  = $(PLATFORM_DEVICE)
-PLATFORM_XDC_TARGET_dm6446_al   = gnu.targets.arm.GCArmv5T
+PLATFORM_XDC_TARGET_dm6446_al   = gnu.targets.UCArm9
 PLATFORM_XDC_dm6446_al      = ti.platforms.evmDM6446
 
 #------------------------------------------------------------------------------
