--- a/dsplink/gpp/src/api/Linux/drv_api.c	2011-05-26 13:59:21.000000000 -0400
+++ b/dsplink/gpp/src/api/Linux/drv_api.c	2010-05-12 14:40:35.000000000 -0400
@@ -469,7 +469,8 @@
         status = DSP_EPOINTER ;
         SET_FAILURE_REASON ;
     }
-
+		
+		DSPLINK_atExitHandler();
     TRC_1LEAVE ("DRV_Finalize", status) ;
 
     return status ;
@@ -1598,7 +1599,7 @@
 
     TRC_1ENTER ("DRV_installCleanupRoutines", linkCfgPtr) ;
 
-    if (linkCfgPtr->gppObject->gppOsObject->handleSignals == TRUE) {
+    if (0 && linkCfgPtr->gppObject->gppOsObject->handleSignals == TRUE) {
         sigemptyset (&block_mask) ;
         sigfillset( &block_mask) ;
         sigInstaller.sa_flags   = (SA_RESTART) ;
@@ -1611,7 +1612,7 @@
                        NULL) ;
         }
     }
-    atexit (&DSPLINK_atExitHandler) ;
+    //atexit (&DSPLINK_atExitHandler) ;
 
     TRC_0LEAVE ("DRV_installCleanupRoutines") ;
 }
