Fix 32pixel alignment of buffer when using accelFrameCopy
Index: gstreamer-ti-svn-r611/src/gsttidmaivideosink.c
===================================================================
--- gstreamer-ti-svn-r611.orig/src/gsttidmaivideosink.c	2012-04-03 11:01:57.789798904 -0700
+++ gstreamer-ti-svn-r611/src/gsttidmaivideosink.c	2012-04-03 11:02:18.305901039 -0700
@@ -1546,7 +1546,7 @@
                          sink->dGfxAttrs.dim.width, dim.width);
                 dim.x = 0;
             } else {
-                dim.x     = ((dim.width - sink->dGfxAttrs.dim.width) / 2) & ~1;
+                dim.x     = ((dim.width - sink->dGfxAttrs.dim.width) / 2) & ~0x1f; // align to 32 pixels 
                 dim.width = sink->dGfxAttrs.dim.width;
             }
 
