--- a/src/gsttividenc1.c
+++ b/src/gsttividenc1.c
@@ -86,6 +86,7 @@ enum
   PROP_GEN_TIMESTAMPS,  /* genTimeStamps  (boolean) */
   PROP_RATE_CTRL_PRESET,/* rateControlPreset  (gint) */
   PROP_ENCODING_PRESET, /* encodingPreset  (gint) */
+  PROP_MAX_IFRAME_INT,  /* maxInterFrameInterval (gint) */
   PROP_BYTE_STREAM      /* byteStream      (gboolean) */
 
 };
@@ -369,6 +370,15 @@ static void gst_tividenc1_class_init(Gst
         g_param_spec_boolean("genTimeStamps", "Generate Time Stamps",
             "Set timestamps on output buffers",
             TRUE, G_PARAM_WRITABLE));
+
+    g_object_class_install_property(gobject_class, PROP_MAX_IFRAME_INT,
+        g_param_spec_int("maxInterFrameInterval",
+            "Max I-Frame Interval",
+            "I to P frame distance"
+            "\n\t\t\t1 - no B frames"
+            "\n\t\t\t2 - one B frame"
+            "\n\t\t\t3 - two B frames",
+            1, G_MAXINT32, 1, G_PARAM_WRITABLE));
 }
 
 /******************************************************************************
@@ -582,6 +592,11 @@ static void gst_tividenc1_set_property(G
             GST_LOG("setting \"genTimeStamps\" to \"%s\"\n",
                 videnc1->genTimeStamps ? "TRUE" : "FALSE");
             break;
+        case PROP_MAX_IFRAME_INT:
+            videnc1->maxInterFrameInterval = g_value_get_int(value);
+            GST_LOG("setting \"maxInterFrameInterval\" to \"%d\" \n",
+                     videnc1->maxInterFrameInterval);
+            break;
         default:
             G_OBJECT_WARN_INVALID_PROPERTY_ID(object, prop_id, pspec);
             break;
--- a/src/gsttividenc1.h
+++ b/src/gsttividenc1.h
@@ -78,6 +78,7 @@ struct _GstTIVidenc1
   gint32         bitRate;
   gint           rateControlPreset;
   gint           encodingPreset;
+  gint           maxInterFrameInterval;
 
   /* Element state */
   Engine_Handle    hEngine;
